openapi/nexus.json

```diff
@@ -2378,6 +2378,62 @@
         }
       }
     },
+    "/v1/instances/{instance}/resize": {
```

Comment:
Is there a specific reason the resize action should be separated from other updateable fields (e.g. `name`) into it's own endpoint? 

Other endpoints generally follow a pattern for "update"  where all updateable fields of a resource are gathered into a single endpoint (e.g. [`VpcUpdate`](https://github.com/oxidecomputer/omicron/blob/b08cce78804c94e19a8404d58beafa3d6d296d4a/openapi/nexus.json#L9513-L9566)  and [`ProjectUpdate`](https://github.com/oxidecomputer/omicron/blob/b08cce78804c94e19a8404d58beafa3d6d296d4a/openapi/nexus.json#L3699-L3744)) for consistency. This pattern makes it a bit easier to avoid endpoint explosion to keep the number of endpoints as small as possible and easier for our users to grok.

Would it be possible to change this endpoint to follow that pattern?

Reply:
There’s precedent here if you think of resize as an instance action like start or stop:

https://github.com/oxidecomputer/omicron/blob/9d8078bc34a78778a3a7d82a58f90a1fe55677df/nexus/tests/output/nexus_tags.txt#L64-L65

Reply:
I think we could have a single instance-update endpoint, but we'd have to decide what to do about the distinction between fields that you can theoretically update on a running instance (name, description) vs. the ones that require the instance to be halted (CPUs, memory, hostname, public keys, user data, etc.). My initial inclination would be to say that all updates (even the "safe" ones) require the instance to be halted, and if it turns out that that's a huge headache for users, we can pivot to allowing updates on running instances if they only touch safe-when-running fields. (This is more complicated to do on the backend, but I don't think it's intractable.) WDYT?

Reply:
Perhaps those two endpoints should be rolled up into a single update endpoint like the others? 🤔 

It may be confusing for users to have some resources with endpoints that take an "update object" and others that have several endpoints to update different parts of a resource. It'd be nice to have some sort of consistency when possible. What do you think?

Reply:
Sorry @gjcolombo! Looks like we were typing at the same time and my answer above looks a bit weird now 😄 

> I think we could have a single instance-update endpoint, but we'd have to decide what to do about the distinction between fields that you can theoretically update on a running instance (name, description) vs. the ones that require the instance to be halted (CPUs, memory, hostname, public keys, user data, etc.). My initial inclination would be to say that all updates (even the "safe" ones) require the instance to be halted, and if it turns out that that's a huge headache for users, we can pivot to allowing updates on running instances if they only touch safe-when-running fields. (This is more complicated to do on the backend, but I don't think it's intractable.) WDYT?

Yeah! I think this is a good compromise. I don't see users changing the name or description of an instance often anyway

Reply:
I could see that working for the fields Greg lists, but not for start and stop and reboot. The problem with the latter is that they take time to complete asynchronously, with intermediate states (`starting`, `stopping`) before the final state, so the PUT can’t return you the resource in a state matching what you sent it (`running`, for example), which is how PUTs are supposed to work.

Regarding requiring the instance to be stopped for all updates, whether the field really needs it or not, I agree that’s fine and easier to understand as a user.

Reply:
> I could see that working for the fields Greg lists, but not for start and stop and reboot. The problem with the latter is that they take time to complete asynchronously, with intermediate states (starting, stopping) before the final state, so the PUT can’t return you the resource in a state matching what you sent it (running, for example), which is how PUTs are supposed to work.

That makes sense, I'm OK with keeping start/stop separate then